language/compiler/ir_to_bytecode/syntax/src/syntax.rs

```diff
@@ -1 +1,1798 @@
+    let s = (tokens.token.1).1;
+    tokens.advance()?;
+    // The token includes a "<" at the end, so chop that off to get the name.
+    Ok(s[..s.len() - 1].to_string())
+}
+
+fn parse_dot_name<'input, 'builder>(tokens: &mut __Matcher<'input, 'builder>)
+    -> Result<String, ParseError<usize, Token<'input>, failure::Error>>
+{
+    if (tokens.token.1).0 != Tok::DotNameValue {
+        return Err(ParseError::InvalidToken { location: tokens.token.0 });
+    }
+    let name = (tokens.token.1).1.to_string();
+    tokens.advance()?;
+    Ok(name)
+}
+
+// AccountAddress: AccountAddress = {
+//     < s: r"0[xX][0-9a-fA-F]+" > => { ... }
+// };
+
+fn parse_account_address<'input, 'builder>(tokens: &mut __Matcher<'input, 'builder>)
+    -> Result<AccountAddress, ParseError<usize, Token<'input>, failure::Error>>
+{
+    if (tokens.token.1).0 != Tok::AccountAddressValue {
+        return Err(ParseError::InvalidToken { location: tokens.token.0 });
+    }
+    let addr = AccountAddress::from_hex_literal(&(tokens.token.1).1)
+        .unwrap_or_else(|_| panic!("The address {:?} is of invalid length. Addresses are at most 32-bytes long", (tokens.token.1).1));
+    tokens.advance()?;
+    Ok(addr)
+}
+
+// Var: Var = {
+//     <n:Name> =>? Var::parse(n),
+// };
+
+fn parse_var<'input, 'builder>(tokens: &mut __Matcher<'input, 'builder>)
+    -> Result<Var, ParseError<usize, Token<'input>, failure::Error>>
+{
+    Var::parse(parse_name(tokens)?)
+}
+
+fn parse_var_<'input, 'builder>(tokens: &mut __Matcher<'input, 'builder>)
+    -> Result<Var_, ParseError<usize, Token<'input>, failure::Error>>
+{
+    let v_start_loc = tokens.token.0;
+    let var = parse_var(tokens)?;
+    let v_end_loc = tokens.token.0;
+    Ok(Spanned{span: Span::new(ByteIndex(v_start_loc as u32),
+                               ByteIndex(v_end_loc as u32)),
+               value: var })
+}
+
+// Field: Field = {
+//     <n:Name> =>? parse_field(n),
+// };
+
+fn parse_field_<'input, 'builder>(tokens: &mut __Matcher<'input, 'builder>)
+    -> Result<Field_, ParseError<usize, Token<'input>, failure::Error>>
+{
+    let start_loc = tokens.token.0;
+    let f = parse_field(parse_name(tokens)?)?;
+    let end_loc = tokens.token.0;
+    Ok(Spanned{span: Span::new(ByteIndex(start_loc as u32),
+                               ByteIndex(end_loc as u32)),
+               value: f })
+}
+
+// CopyableVal: CopyableVal = {
+//     AccountAddress => CopyableVal::Address(<>),
+//     "true" => CopyableVal::Bool(true),
+//     "false" => CopyableVal::Bool(false),
+//     <i: U64> => CopyableVal::U64(i),
+//     <buf: ByteArray> => CopyableVal::ByteArray(buf),
+// }
+
+fn parse_copyable_val_<'input, 'builder>(tokens: &mut __Matcher<'input, 'builder>)
+    -> Result<CopyableVal_, ParseError<usize, Token<'input>, failure::Error>>
+{
+    let start_loc = tokens.token.0;
+    let val = match (tokens.token.1).0 {
+        Tok::AccountAddressValue => {
+            let addr = parse_account_address(tokens)?;
```

Comment:
I think if you move the `tokens.advance()` outside of `parse_account_address`, you can simplify this match by doing `tokens.advance()?` only once at line 213 instead of in each branch.

Reply:
I think you're right, but I'm not sure that's a good idea. I tried to maintain a convention that each "parse" function consumes that tokens that it processes. Making an exception for parse_account_address means that the call in parse_qualified_module_ident would need to be adjusted to skip over the account address token. Keeping the consistent convention seems more important than moving tokens.advance() calls out of that match. Are you OK with that?

Reply:
Yep, that makes sense. Not worth it to make the code smaller.